### ISSUE TITLE
[DISCUSSION] Grid-box corner coordinates in (or appended to) diagnostics

### ISSUE THREAD

**Issue by LiamBindle** [geoschem/GCHP#38]
To plot data on a curvlinear grid with routines like matplotlib's `pcolormesh()`, the coordinates of grid-box edges are necessary. Currently the diagnostics don't include edge coordinates, and there's no easy way to get them.

### GCPy calculates edge coordiantes internally
GCPy calculates edge coordinates itself (privately), and uses those to plot cubed-sphere data. (See [here](https://github.com/geoschem/gcpy/blob/1fa302b66842c541b69984bc4a0f3fb8eb7658dc/gcpy/plot.py#L2596-L2617)).
```python
    else:
        #Cubed-sphere single level
        ax.coastlines()
        try:
            if masked_data == None:
                masked_data = np.ma.masked_where(np.abs(grid["lon"] - 180) < 2, plot_vals.data.reshape(6, res, res))
        except ValueError:
            #Comparison of numpy arrays throws errors
            pass
        [minlon,maxlon,minlat,maxlat] = extent
        #Catch issue with plots extending into both the western and eastern hemisphere
        if np.max(grid["lon_b"] > 180):
            grid["lon_b"] = (((grid["lon_b"]+180)%360)-180)
        for j in range(6):
            plot = ax.pcolormesh(
                grid["lon_b"][j, :, :],
                grid["lat_b"][j, :, :],
                masked_data[j, :, :],
                transform=proj,
                cmap=comap,
                norm=norm
            )
```
In this snippet, the `grid` dict contains arrays with edge coordinates. The grid dict is generated by `call_make_grid()`. But, these coordinates are not available to users.

### Potential solutions
Below are some ideas that come to mind:
1. Add a command-line tool to GCPy that appends the edge coordinates (shape: nf, YEdim, XEdim; where YEdim, XEdim are csres+1 in size) to a give file
2. Perhaps there's a way to add this information to the diagnostics with MAPL?

I look forward to the thoughts of others. Is there a solution to this that I'm not aware of?

**Comment by sdeastham** [comment]
I agree this would be great. I'm guessing there *is* a way to get this data out of HISTORY, but would want to ask the MAPL folks about it, as the collection would be a weird shape.

One possibility is to make GCPy explicitly require a grid specification, but add tools which can estimate it from lons and lats in a NetCDF file. That would perhaps be asking too much of the user though? I have routines to do this in github.com/sdeastham/gcgridobj but that's because `gcgridobj` is basically designed exactly for my work flow =/

**Comment by lizziel** [comment]
Are the edge coordinates already stored in State_Grid in GCHP? If yes, we could get them out via History that way. Not as nice as having them in every file ready for use but it would be better than estimating.

I am bringing in @tclune on whether this is something GMAO has considered for MAPL.

**Comment by tclune** [comment]
I have not ever seen a request for edge coordinates.      We have centers and corners.   In theory edges are derivable from the corners.     If this is important, we could look at adding this as metadata for grids.  If done correctly the computational cost ought to be tolerably low.

**Comment by LiamBindle** [comment]
@tclune Sorry, I meant corner coordaintes. That would be terrific if we could get those from MAPL. How would we do this?

**Comment by tclune** [comment]
OK - let me dig a bit.   I may be confusing myself in terms of what MAPL has already - esp. now that we are using ESMF for the CS.   From an ESMF perspective, the corners are weird as they do not duplicate them on the face edges/corners.   I.e. they don't divide neatly into an array that is `6 x NX x NX` the way that the centers do.    If you are ok with duplicated corners, such that they can be stored in an array that is `6 x (NX+1) x (NX+1)` I think this will be straightforward.   Maybe a tiny bit of tedious code to fill in the points that ESMF does not duplicate.

**Comment by LiamBindle** [comment]
@tclune Duplicate corners would be preferred actually since routines like `pcolormesh()` expect `X,Y` to be size `N+1,N+1`

**Comment by tclune** [comment]
I have discussed with @atrayanov and @ben-nasa  and they think this will be straightforward.      Please open a ticket in MAPL and assign to @ben-nasa.     

Basically we already have a per-grid factory with a method that specifies metadata.   That routine would need a few more lines to add metadata for corners.   And then the generic layer that sends data to the I/O server needs a few lines to extract the actual values of the coordinates from the ESMF Grid object.   The biggest concern is the handling of the duplicate corners, and the MAPL ticket you would open is the best place to discuss that.

**Comment by LiamBindle** [comment]
@lizziel This will be an update in MAPL. Will this have to come into GCHP post 13.0.0 then? If so, could cherry-pick it so it gets into 13.0.0?

**Comment by lizziel** [comment]
I'd prefer to have an official MAPL version in the 13.0.0 release. @sdeastham, do you have an opinion on this?

**Comment by tclune** [comment]
At this time I do not want to commit when this would go into a release of MAPL.   We will likely need to do an early 2.3 release because of some smaller features to support GOCART in NOAA.  (I.e. the big features that were aimed in 2.3 will probably slip into 2.4).   

So much depends on when you expect to need a "final" MAPL for 13.0.0

**Comment by lizziel** [comment]
I was not planning on updating MAPL beyond 2.2.7 which I currently have working. We plan to start benchmarking GEOS-Chem in only a few weeks and I do not want to jeopardize the schedule given we are already months behind.

**Comment by lizziel** [comment]
Corner coordinates are included in output files starting in GCHP 13.1. That version upgrades MAPL from v2.2.7 to v2.6.3. The new variables and dimensions appear in the file as follows: 

```
dimensions:
        XCdim = 25 ;
        Xdim = 24 ;
        YCdim = 25 ;
        Ydim = 24 ;
        lev = 72 ;
        ncontact = 4 ;
        nf = 6 ;
        orientationStrLen = 5 ;
        time = UNLIMITED ; // (1 currently)
variables:
        float corner_lons(nf, YCdim, XCdim) ;
                corner_lons:long_name = "longitude" ;
                corner_lons:units = "degrees_east" ;
        float corner_lats(nf, YCdim, XCdim) ;
                corner_lats:long_name = "latitude" ;
                corner_lats:units = "degrees_north" ;
```
